src/connection.ts

```diff
@@ -4,70 +4,235 @@ import { AbortOptions } from '@libp2p/interfaces';
+
+      log.trace(`${logPrefix} supported protocols - ${protocols}`);
+
+      let { stream, protocol } = await mshandle(rawStream, protocols, { signal: controller.signal });
+      if (metrics) {
+        metrics.trackStream({ stream, protocol, remotePeer: this.remotePeer });
+      }
+
+      log.trace(`${logPrefix} handled protocol - ${protocol}`);
+
+      rawStream.stat.protocol = protocol;
+      let result = this.wrapMsStream(rawStream, stream);
+
+      this.addStream(result);
+
+      // handle stream
+      let { handler } = registrar.getHandler(protocol);
+      handler({ connection: this, stream: result });
     };
-    this.remotePeer = init.remotePeerId;
-    // for muxing incoming stream
-    // this._peerConnection.ondatachannel = ({ channel }) => {
-    // 	let stream = DataChannelStream(channel)
-    // 	this.addStream(stream)
-    // }
   }
 
-  async newStream(multicodecs: string | string[], options?: AbortOptions): Promise<ic.Stream> {
-    // let label = uuid.v4()
-    // let dc = this._peerConnection.createDataChannel(label, {})
-    // await datachannel opening
-    // return DataChannelStream(dc)
-    log('TODO', this.ufrag);
-    this.peerConnection.createDataChannel(genUuid());
-    throw new Error('not implemented');
+  private wrapMsStream(rawStream: WebRTCStream, stream: Duplex<Uint8ArrayList, Uint8ArrayList | Uint8Array, Promise<void>>): ic.Stream {
+    return {
+      ...stream,
+      close: () => {
+        rawStream.close();
+      },
+      closeRead: () => {
+        rawStream.closeRead();
+      },
+      closeWrite: () => {
+        rawStream.closeWrite();
+      },
+      abort: (err) => {
+        rawStream.abort(err);
+      },
+      reset: () => rawStream.reset(),
+      id: rawStream.id,
+      metadata: rawStream.metadata,
+      stat: rawStream.stat,
+    };
+  }
+
+  private findStreamLimit(protocol: string, direction: ic.Direction): number {
+    let registrar = this.components.getRegistrar();
+    try {
+      let handler = registrar.getHandler(protocol);
+      return direction === 'inbound' ? handler.options.maxInboundStreams || DEFAULT_MAX_INBOUND_STREAMS : handler.options.maxOutboundStreams || DEFAULT_MAX_OUTBOUND_STREAMS;
+    } catch (err) {}
+    return direction === 'inbound' ? DEFAULT_MAX_INBOUND_STREAMS : DEFAULT_MAX_OUTBOUND_STREAMS;
+  }
+
+  private countStream(protocol: string, direction: ic.Direction): number {
+    return this.streams.filter((s) => s.stat.protocol === protocol && s.stat.direction === direction).length;
+  }
+
+  async newStream(protocols: string | string[], options: AbortOptions = {}): Promise<ic.Stream> {
+    let label = genUuid().slice(0, 8);
+    let openPromise = defer();
+    let abortedPromise = defer();
+    let controller: TimeoutController | undefined;
+    let metrics = this.components.getMetrics();
+    let openError: Error | undefined;
+
+    log.trace(`opening new stream with protocols: ${protocols}`);
```

Comment:
I see that we're not testing stream limits here. So those limits _only_ apply to incoming streams, not the sum of the two?